src/internal/connector/onedrive/drive.go

```diff
@@ -7,101 +7,117 @@ import (
 	"github.com/alcionai/corso/src/pkg/logger"
 )
 
 var errFolderNotFound = errors.New("folder not found")
 
 const (
+	getDrivesRetries = 3
+
 	// nextLinkKey is used to find the next link in a paged
 	// graph response
-	nextLinkKey           = "@odata.nextLink"
-	itemChildrenRawURLFmt = "https://graph.microsoft.com/v1.0/drives/%s/items/%s/children"
-	itemByPathRawURLFmt   = "https://graph.microsoft.com/v1.0/drives/%s/items/%s:/%s"
-	itemNotFoundErrorCode = "itemNotFound"
-	userMysiteURLNotFound = "BadRequest Unable to retrieve user's mysite URL"
-	userMysiteNotFound    = "ResourceNotFound User's mysite not found"
+	nextLinkKey             = "@odata.nextLink"
+	itemChildrenRawURLFmt   = "https://graph.microsoft.com/v1.0/drives/%s/items/%s/children"
+	itemByPathRawURLFmt     = "https://graph.microsoft.com/v1.0/drives/%s/items/%s:/%s"
+	itemNotFoundErrorCode   = "itemNotFound"
+	userMysiteURLNotFound   = "BadRequest Unable to retrieve user's mysite URL"
+	userMysiteNotFound      = "ResourceNotFound User's mysite not found"
+	contextDeadlineExceeded = "context deadline exceeded"
 )
 
-// Enumerates the drives for the specified user
-func drives(
-	ctx context.Context,
-	service graph.Servicer,
-	resourceOwner string,
+type drivePager interface {
+	GetPage(context.Context) (api.PageLinker, error)
+	SetNext(nextLink string)
+	ValuesIn(api.PageLinker) ([]models.Driveable, error)
+}
+
+func PagerForSource(
 	source driveSource,
-) ([]models.Driveable, error) {
+	servicer graph.Servicer,
+	resourceOwner string,
+	fields []string,
+) (drivePager, error) {
 	switch source {
 	case OneDriveSource:
-		return userDrives(ctx, service, resourceOwner)
+		return api.NewUserDrivePager(servicer, resourceOwner, fields), nil
 	case SharePointSource:
-		return siteDrives(ctx, service, resourceOwner)
+		return api.NewSiteDrivePager(servicer, resourceOwner, fields), nil
 	default:
 		return nil, errors.Errorf("unrecognized drive data source")
 	}
 }
 
-func siteDrives(ctx context.Context, service graph.Servicer, site string) ([]models.Driveable, error) {
-	options := &sites.ItemDrivesRequestBuilderGetRequestConfiguration{
-		QueryParameters: &sites.ItemDrivesRequestBuilderGetQueryParameters{
-			Select: []string{"id", "name", "weburl", "system"},
-		},
-	}
+func drives(
+	ctx context.Context,
+	pager drivePager,
+	retry bool,
+) ([]models.Driveable, error) {
+	var (
+		numberOfRetries = getDrivesRetries
+		drives          = []models.Driveable{}
+	)
 
-	r, err := service.Client().SitesById(site).Drives().Get(ctx, options)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to retrieve site drives. site: %s, details: %s",
-			site, support.ConnectorStackErrorTrace(err))
+	if !retry {
+		numberOfRetries = 0
 	}
 
-	return r.GetValue(), nil
-}
+	for done := false; !done; {
+		// Retry Loop for Drive retrieval. Request can timeout
+		for i := 0; i <= numberOfRetries; i++ {
+			page, err := pager.GetPage(ctx)
+			if err == nil {
```

Comment:
I think this would flow better if `err != nil` was handled here instead of the success path.  I'm not seeing anything that requires the success handling to go first, so the inversion does weird things for my golang-standards expectations.  Likewise, it feels a bit strange that the happy path "exit point" is a break at the very center of the code block, as opposed to the end of it.